core/arch/mangle_shared.c

```diff
@@ -272,24 +279,38 @@ prepare_for_clean_call(dcontext_t *dcontext, clean_call_info_t *cci, instrlist_t
 
     /* check if need adjust stack for alignment. */
     if (cci->should_align) {
-#if (defined(X86) && defined(X64)) || defined(MACOS)
+#if defined(X86) && (defined(X64) || defined(UNIX))
         /* PR 218790: maintain 16-byte rsp alignment.
          * insert_parameter_preparation() currently assumes we leave rsp aligned.
          */
+        int align = get_ABI_stack_alignment();
+        int off = align - (dstack_offs % align);
+        ASSERT(off % XSP_SZ == 0);
+        /* Make sure cleanup_after_clean_call() can compute the same offset.
+         * We could make the caller pass back in dstack_offs except for
+         * dr_cleanup_after_call().
+         */
+        int simd = 0;
+        if (clean_call_needs_simd(cci)) {
+            simd = MCXT_TOTAL_SIMD_SLOTS_SIZE + MCXT_TOTAL_OPMASK_SLOTS_SIZE +
+                PRE_XMM_PADDING;
+        }
         uint num_slots = DR_NUM_GPR_REGS + NUM_EXTRA_SLOTS;
         if (cci->skip_save_flags)
             num_slots -= 2;
-        num_slots -= cci->num_regs_skip; /* regs that not saved */
+#    ifndef X86_32 /* x86 uses pusha regardless of regs we could skip */
+        num_slots -= cci->num_regs_skip; /* regs not saved */
+#    endif
+        IF_DEBUG(uint emulate_dstack_offs = simd + num_slots * XSP_SZ;)
+        ASSERT(emulate_dstack_offs == dstack_offs || cci->out_of_line_swap);
```

Comment:
Nit: maybe place the calculation of emulate_dstack_offs, including the num_slot counts, inside a helper function? The counting of num_slots here is only used for debugging (namely the for the ASSERT) because insert_push_all_registers() also does the calculation. Calling a function will make it easier for you to ifdef this code  out and only be executed by debug builds. Plus, you will avoid repetition of code at line 336.

Reply:
+1, done.